Develop aircraft: 1) Vertical profile: Option for Shading or Box  IQR (Box case with whiskers high and low: 90th and 10th percentile) with median profiles, 2) Time series with altitude profile as secondary axis (example case: FIREX vs WRF-Chem)

quaz115 opened the pull request:
This is still in works, but close to completion (median and interquartile shaded regions seem to be plotted fine). @colin-harkins suggested to create a draft PR for now (need feedback so opening for comments for now, can switch back to draft PR as needed)

 NOTE: aircraftplots.py is just a copy of surfplots.py, just with additional 'make_vertprofile' function 
Used similar plotting template as 'make_timeseries', just swapped  x and y coordinates in vertical profile (in function 'make_vertprofile' in new aircraftplots.py, Lines 503-660)

@rschwant @colin-harkins Check current vertprofile output plots on Meiyu at: 
/wrk/qrasool/output_testaircraftrhs/
          plot_grp2.vertprofile.NO2_CL_RYERSON.2019-09-05_12.2019-09-06_00.all.CONUS.png
          plot_grp2.vertprofile.NO_CL_RYERSON.2019-09-05_12.2019-09-06_00.all.CONUS.png
          plot_grp2.vertprofile.O3_CL_RYERSON.2019-09-05_12.2019-09-06_00.all.CONUS.png
    
 
**Need to resolve some plotting issues:** (Lines 601-602, 631-632 in aircraftplots.py)
a)  **plot_kwargs or **plot_dict not working currently in 'make_vertprofile' (i.e. vertical profile of model vs aircraft observations)  (had to comment out currently)
 b) Legends like time-series plot: firexaq (aircraft obs) and wrfchem_v4.2 (model) not working
Currently, it plots the vertical profile but with (following warnings):
`_No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument._`

zmoon commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,1241 @@
+# Copyright (C) 2022 National Center for Atmospheric Research and National Oceanic and Atmospheric Administration
+# SPDX-License-Identifier: Apache-2.0
+#
+#Code to create plots for surface observations
```
I seem some functions in here that I believe are (also) defined elsewhere in MM. Could they be imported instead?

quaz115 replied:
@zmoon Can you provide any suggested template/example I should follow to achieve that?

zmoon replied:
For example, if you want `new_color_map` from `surfplots` you can use
```python
from .surfplots import new_color_map
```

quaz115 commented:
@colin-harkins @zmoon :  Also, the latest commit [https://github.com/NOAA-CSL/MELODIES-MONET/pull/181/commits/b0e57f764a2faafbe94fbb57a2ce5292aeabefc4] works for having vertical profile plots either as: a) shading (interquartile range) or b) box (interquartile range)-whisker (10th-90th percentiles) style, with median lines. Examples attached as follows: 
a) interquartile_Style == 'shading' (interquartile range): ![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/45dca235-13c5-4b0e-ac47-e2528fbe1a97)

b) interquartile_style == 'box' i.e., Box-whisker (box: interquartile range, whisker: low- 10th percentile, high- 90th percentile):  
![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/a3191fab-fd76-437a-9b4d-0f1b1a130a69)

quaz115 commented:
Another minor [update](https://github.com/NOAA-CSL/MELODIES-MONET/pull/181/commits/372d3ce5622f465fa85f1e1790c0b15d3e5b7ab6) to have y-axis label (Altitude) in vertical profile as integers   
![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/107c8fcd-3cef-4069-aca4-f632323d5c74)

quaz115 commented:
> @colin-harkins you can review these changes in the latest commit to have an Altitude secondary axis for time_series in case of develop_aircraft (i.e. if altitude_variable is present), adding the resultant plot example generated: [2f5cdec](https://github.com/NOAA-CSL/MELODIES-MONET/commit/2f5cdec2d43f8ed21f594634aa1803a5b6750ee7) ![image](https://user-images.githubusercontent.com/63623865/259184241-46a22c06-96ec-479e-a408-320e8067eb0d.png)
@colin-harkins Now, 'add_yax2_altitude()' is a separate function in aircraftplots.py and is simply called in driver.py. See this [commit](https://github.com/NOAA-CSL/MELODIES-MONET/pull/181/commits/dd6a9ebba1d5d9d17efd143f1cf39be5e669ce34) for details of the change.
![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/2c0cedda-8041-4afc-8076-9739a29d3bb6)

quaz115 commented:
Added a subset option for aircraft time-series (with altitude) based on limits of altitude (secondary y-axis specified, i.e. vmin2 and vmax2: these are optional in yaml configuration and default plots the full data frame). Also, I updated the docstring for 'add_y2axis_altitude()' in aircraftplots.py for more description. See  the [commit](https://github.com/NOAA-CSL/MELODIES-MONET/pull/181/commits/a9d893dc75e86f9dcc1c4072dfed79e197bbe3d9) for the changes/updates

![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/01700033-87f7-4ebe-84a3-2016b6436cdb)
![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/652de36b-8d47-4586-b55e-d7eebddbc56a)
Full data frame time-series with altitude profile plot (ozone, just for reference):  
![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/0571447d-8ee1-4c3a-8323-967fc40190ac)


Example of vertical profile as a subset of y-axis limits defined by vmin, vmax in yaml (optional as well, and plots full data frame if not specified in yaml)
![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/4aec067e-12ed-4e27-8d28-bf003568ef36)
![image](https://github.com/NOAA-CSL/MELODIES-MONET/assets/63623865/9234a830-f39e-4959-a64e-2f712deeeb01)

colin-harkins commented on examples/jupyter_notebooks/Monet-analysis-example-plots-wrf_AircraftVertprofileTry1_QZR.ipynb:
I think that this needs to be updated as it doesn't use the yaml file that has been added to examples/yaml

colin-harkins commented on examples/jupyter_notebooks/Monet-analysis-example-plots-wrf_AircraftVertprofileTry1_QZR.py:
I think that this needs to be updated as it doesn't use the yaml file that has been added to examples/yaml. Also potentially remove the file since it's not a notebook.

colin-harkins commented on examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft.yaml:
```diff
@@ -0,0 +1,179 @@
+    domain_name: ['CONUS'] #List of domain names. If domain_type = all domain_name is used in plot title.
+    data: ['firexaq_wrfchem_v4.2'] # make this a list of pairs in obs_model where the obs is the obs label and model is the model_label
+    data_proc:
+      altitude_variable: altitude  
+      altitude_ticks: 1000  # Altitude tick interval in meters (for secondary y-axis for altitude (m))
```
not necessary right now but in the future it might make more sense to contain all the altitude relevant variables in a list like altitude_yax2, which would contain everything like vmin2, vmax2, altitude_variable, altitude_ticks, ylabel2, plot_kwargs_y2. This would also make abstracting this to generally plot any second axis variable very easy.

colin-harkins commented on examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft.yaml:
```diff
@@ -0,0 +1,179 @@
+      rem_obs_nan: True  # True: Remove all points where model or obs variable is NaN. False: Remove only points where model variable is NaN.
+      set_axis: False  # If set to True, add vmin_plot and vmax_plot for each variable in obs.
+      interquartile_style: 'shading' # or 'box'
+    #altitude_variable: 'Static_Pressure_YANG' or 'pressure_obs' (=P_BUI(hPa)*100) (pressure in Pa)  # Specify the name of the altitude variable (pressure or MSL height) 
+    altitude_variable: altitude #'MSL_GPS_Altitude_YANG' #'ALTITUDE' in m or 
```
This suggests that pressure can also be plotted but the labeling in make_vertprofile() is currently all specific to altitude in meters. Would be nice to add in plotting by pressure in the future, but it would require some additional coding such as flipping the y axis to start from high values and decrease as it goes up. For now I would remove the commented line and say that altitude must be in meters after unit_scale application on the variable.

colin-harkins commented on melodies_monet/driver.py:
```diff
@@ -946,11 +947,28 @@ def plotting(self):
+            # Get the plot type
             plot_type = grp_dict['type']
 
+        
+        ##for grp, grp_dict in plot_dict.items():
+            ##pair_labels = grp_dict['data']
+            # get the plot type
+            ##plot_type = grp_dict['type']
+
+          
+
```
```suggestion
```

colin-harkins commented on melodies_monet/driver.py:
```diff
@@ -1156,7 +1176,22 @@ def plotting(self):
+                            else:
+                                a_w = None
+
+                            #Steps needed to subset if secondary y-axis (altitude_variable) limits are provided, as make_timeseries from surfaceplots.py already plots the whole df
+                            # Extract altitude (secondary y-axis) limits (if provided)
+                            vmin2 = grp_dict['data_proc'].get('vmin2', None)
+                            vmax2 = grp_dict['data_proc'].get('vmax2', None)
+                            # Subset the data based on vmin2 and vmax2 if provided
+                            if vmin2 is not None and vmax2 is not None:
+                                altitude_variable = grp_dict['data_proc']['altitude_variable']
+                                pairdf = pairdf[(pairdf[altitude_variable] >= vmin2) & (pairdf[altitude_variable] <= vmax2)]
+
```
I think that the use of vmin2 and vmax2 should be consistent with vmin_plot and vmax_plot for the primary axis, where it uses vmin_plot and vmax_plot to set the ylim of the primary axis. The same subsetting functionality here can be achieved with either the filter_string and filter_dict yaml options and it could be confusing to the user if the functionality of vmin and vmax is different throughout. Also setting altitude_variable here is redundant because it is set below.

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
For now I would probably just remove aircraftplots.make_spatial_overlay(), aircraftplots.make_spatial_bias(), and aircraftplots.make_spatial_bias_exceedance() or just delete the code within them here so that there isn't any confusion over whether they are implemented.

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,842 @@
+        Matplotlib ax such that driver.py can iterate to overlay multiple models on the same plot.
+    """
+    ax2 = ax.twinx()
+    ax2.plot(pairdf.index, pairdf[altitude_variable], color='g', label='Altitude (m)')
```
Not necessary right now but in the future it would be good to allow the color, linestyle and markers of the altitude axis to be set in the control file. Also I would potentially default to having some kind of marker because the obs and model variables the lack of a marker indicates missing timeseries data, which is important. Here the line just connects between any missing data points and can make it seem like the altitude follows that connection line rather than the data just being missing.

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,842 @@
+    ax2.plot(pairdf.index, pairdf[altitude_variable], color='g', label='Altitude (m)')
+    ax2.set_ylabel('Altitude (m)', color='g', fontweight='bold', fontsize=text_kwargs['fontsize'])
+    ax2.tick_params(axis='y', labelcolor='g', labelsize=text_kwargs['fontsize'] * 0.8)
+    ax2.set_ylim(0, max(pairdf[altitude_variable]) + altitude_ticks)
+    ax2.set_xlim(ax.get_xlim())
+    ax2.yaxis.set_ticks(np.arange(0, max(pairdf[altitude_variable]) + altitude_ticks, altitude_ticks))
```
Like my prior comment, I think this should be controlled by `vmin2` and `vmax2`. I think it's potentially fine to use the current functionality as the default but if a user wants to change the secondary yaxis limits directly they should be able to.

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,842 @@
+        # Convert bin_midpoints to a numerical data type
+        df['bin_midpoints'] = df['bin_midpoints'].astype(float)
+
+        p5 = df.groupby(altitude_bins)[column].quantile(0.05)
+        p10 = df.groupby(altitude_bins)[column].quantile(0.10)
+        p90 = df.groupby(altitude_bins)[column].quantile(0.90)
+        p95 = df.groupby(altitude_bins)[column].quantile(0.95)
```
`p5` and `p95` are never used.

```suggestion
        p10 = df.groupby(altitude_bins)[column].quantile(0.10)
        p90 = df.groupby(altitude_bins)[column].quantile(0.90)
```

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,842 @@
+        
+        # Plot shaded interquartiles
+        plot_kwargs_fillbetween['label'] = f"{label} (interquartile 25th-75th percentile range)"
+        if interquartile_style == 'shading':
+            ax.fill_betweenx(binmidpoint.values, q1.values, q3.values, alpha=0.3, **plot_kwargs_fillbetween)
```
Not necessary right now but in the future it would be nice to have the ability to set the range in the shading method in the control file, potentially with it defaulting to 25th to 75th if nothing else set.

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,842 @@
+        plot_kwargs['label'] = f"{label} (median)"
+        median_line_df = pd.DataFrame(data={'median': median.values, 'binmidpoint': binmidpoint.values})
+        ax = median_line_df.plot(x='median', y='binmidpoint', ax=ax, legend=True, **plot_kwargs)
+        #ax= median_line_df.plot(ax=ax, legend=True, **plot_kwargs)
+        
+                
```
```suggestion
```

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,842 @@
+        df['bin_midpoints'] = df['bin_midpoints'].astype(float)
+
+        # Calculate the 10th, 90th, 5th, and 95th percentiles
+        p10 = df.groupby(altitude_bins)[column].quantile(0.10)
+        p90 = df.groupby(altitude_bins)[column].quantile(0.90)
+        p5 = df.groupby(altitude_bins)[column].quantile(0.05)
+        p95 = df.groupby(altitude_bins)[column].quantile(0.95)
```
`p5` and `p95` are never used.
```suggestion
        p10 = df.groupby(altitude_bins)[column].quantile(0.10)
        p90 = df.groupby(altitude_bins)[column].quantile(0.90)
```

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,842 @@
+        ax = median_line_df.plot(x='median', y='binmidpoint', ax=ax, legend=True, **plot_dict)
+        
+    
+    # Add text to legend (adjust the x and y coordinates to place the text below the legend)
+    plt.text(1.12, 0.7, 'Bounds of box: Interquartile range\nWhiskers: 10th and 90th percentiles', transform=ax.transAxes, fontsize=text_kwargs['fontsize']*0.8)
```
```suggestion
    if interquartile_style == 'box':
        # Add text to legend (adjust the x and y coordinates to place the text below the legend)
        plt.text(1.12, 0.7, 'Bounds of box: Interquartile range\nWhiskers: 10th and 90th percentiles', transform=ax.transAxes, fontsize=text_kwargs['fontsize']*0.8)
```

colin-harkins reviewed (commented):
Really great job on this @quaz115! There's just a couple smaller things that I noticed reviewing here but they should be quite easy to fix. Also I noted some things that it would be nice to include in the future. Lets focus on getting this pull request merged into develop_aircraft first and then address those nice to have items in later pull requests.

colin-harkins commented on melodies_monet/plots/aircraftplots.py:
```diff
@@ -0,0 +1,538 @@
+    # Set labels, ticks, and limits
+    ax2.set_ylabel(ylabel2, fontweight='bold', fontsize=text_kwargs['fontsize'], color=plot_kwargs_y2.get('color', 'g'))
+    ax2.tick_params(axis='y', labelcolor=plot_kwargs_y2.get('color', 'g'), labelsize=text_kwargs['fontsize'] * 0.8)
+    ax2.set_ylim(pairdf[altitude_variable].min(), pairdf[altitude_variable].max() + altitude_ticks)
```
Can you make this the default but add the option for vmin_y2 and vmax_y2 in the `altitude_yax2` dict in the control yaml and if they are defined then use those to set the y limits? That provides a slightly different functionality than the filtering, such as if you wanted to have a wider range of altitude plotted than min/max.

colin-harkins reviewed (commented):
One last comment but otherwise this looks great and we can merge it to develop_aircraft once that's addressed. Once it's merged I'll pull in these updates to #185 and resolve any potential merge conflicts.

colin-harkins reviewed (approved):
@quaz115 I think this looks good. You can merge it into the develop_aircraft branch.
